src/channel.rs

```diff
@@ -0,0 +1,284 @@
+#![allow(unused_imports)]
+
+use {io, Evented, EventSet, Poll, PollOpt, Registration, SetReadiness, Token};
+use lazy::{Lazy, AtomicLazy};
+use std::sync::{mpsc, Arc};
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+#[derive(Clone)]
+pub struct SenderCtl {
+    inner: Arc<Inner>,
+}
+
+pub struct ReceiverCtl {
+    registration: Lazy<Registration>,
+    inner: Arc<Inner>,
+}
+
+pub struct Sender<T> {
+    tx: StdSender<T>,
+    ctl: SenderCtl,
+}
+
+enum StdSender<T> {
+    Bounded(mpsc::SyncSender<T>),
+    Unbounded(mpsc::Sender<T>),
+}
+
+pub struct Receiver<T> {
+    rx: mpsc::Receiver<T>,
+    ctl: ReceiverCtl,
+}
+
+#[derive(Debug)]
+pub enum SendError<T> {
+    Io(io::Error),
+    Disconnected(T),
+}
+
+#[derive(Debug)]
+pub enum TrySendError<T> {
+    Io(io::Error),
+    Full(T),
+    Disconnected(T),
+}
+
+struct Inner {
+    pending: AtomicUsize,
+    set_readiness: AtomicLazy<SetReadiness>,
+}
+
+pub fn from_std_channel<T>((tx, rx): (mpsc::Sender<T>, mpsc::Receiver<T>)) -> (Sender<T>, Receiver<T>)
+{
+    let (tx_ctl, rx_ctl) = ctl_pair();
+
+    let tx = Sender {
+        tx: StdSender::Unbounded(tx),
+        ctl: tx_ctl,
+    };
+
+    let rx = Receiver {
+        rx: rx,
+        ctl: rx_ctl,
+    };
+
+    (tx, rx)
+}
+
+pub fn from_std_sync_channel<T>((tx, rx): (mpsc::SyncSender<T>, mpsc::Receiver<T>)) -> (Sender<T>, Receiver<T>)
+{
+    let (tx_ctl, rx_ctl) = ctl_pair();
+
+    let tx = Sender {
+        tx: StdSender::Bounded(tx),
+        ctl: tx_ctl,
+    };
+
+    let rx = Receiver {
+        rx: rx,
+        ctl: rx_ctl,
+    };
+
+    (tx, rx)
+}
+
+pub fn ctl_pair() -> (SenderCtl, ReceiverCtl) {
+    let inner = Arc::new(Inner {
+        pending: AtomicUsize::new(0),
+        set_readiness: AtomicLazy::new(),
+    });
+
+    let tx = SenderCtl {
+        inner: inner.clone(),
+    };
+
+    let rx = ReceiverCtl {
+        registration: Lazy::new(),
+        inner: inner,
+    };
+
+    (tx, rx)
+}
+
+impl SenderCtl {
+    /// Call to track that a message has been sent
```

Comment:
Maybe worth documenting what can be done in the case of an error? E.g. the readiness is not set, but the increment still happened